lib/sshkit/ssh/client_key_api.ex

```diff
@@ -0,0 +1,130 @@
+defmodule SSHKit.SSH.ClientKeyApi do
+
+  @behaviour :ssh_client_key_api
+
+  @moduledoc ~S"""
+
+  Simple wrapper for the Erlang `:ssh_client_key_api` behavior, to
+  to make it easier to specify SSH keys and known_hosts files independently of
+  any particular users home directory
+
+
+  It is meant to primarily be used via the convenience function `with_config`:
+
+  `
+  cb = SSHKit.SSH.ClientKeyApi.with_options(identity: "path/to/keyfile", known_hosts: "path/to/known_hosts", accept_hosts: false)  
+  `
+
+  The result can be passed as an option when creating an `SSHKit.SSH.Connection`:
+
+  `
+  SSHKit.SSH.connect("example.com", key_cb: cb)
+  `
+
+    - `identity`: `IO.device` providing the ssh key (required)
+    - `known_hosts`: `IO.device` providing the known hosts list (required)
+    - `accept_hosts`: `boolean` silently accept and add new hosts to the known hosts. By default only known hosts will be accepted. 
+  """
+
+  @spec with_options(opts :: list) :: {atom, list}
+  @doc """
+    returns a tuple suitable for passing the `SSHKit.SSH.Connect` as the `key_cb` option.
+#### Options
+     - key_file - path to SSH key to use for authentication
+     - known_hosts_file - path to file containing fingerprints of hosts to recognize
+     - accept_hosts - whether or not to silently accept unknown host fingerprints
+
+     by default it will use the the files found in `System.user_home!`
+     
+### Example
+
+      `
+      cb = SSHKit.SSH.ClientKeyApi.with_options(identity: "path/to/keyfile", known_hosts: "path/to/known_hosts", accept_hosts: false)  
+      SSHKit.SSH.connect("example.com", key_cb: cb)
+      `
+
+  """
+  def with_options(opts \\ []) do    
+      opts = with_defaults(opts, default_opts())     
+      opts = 
+      opts
+        |> Keyword.put(:identity, File.read!(opts[:identity]))
+    {__MODULE__, opts}
+  end  
+
+  def add_host_key(hostname, key, opts) do      
+    case accept_hosts(opts) do
+      true -> 
+        opts
+        |> known_hosts
+        |> :public_key.ssh_decode(:known_hosts)
+        |> (fn decoded -> decoded ++ [{key, [{:hostnames, [hostname]}]}] end).()
+        |> :public_key.ssh_encode(:known_hosts)
+        |> (fn encoded -> File.write!(known_hosts_file(opts), encoded) end).()
+      _ -> 
+        message = 
+          """
+          Error: unknown fingerprint found for #{inspect hostname} #{inspect key}.
+          You either need to add a known good fingerprint to your known hosts file for this host,
+          *or* pass the accept_hosts option to your client key callback
+          """        
+        {:error, message}
+    end    
+  end
+
+  def is_host_key(key, hostname, _alg, opts) do    
+    opts
+    |> known_hosts    
+    |> :public_key.ssh_decode(:known_hosts)
+    |> has_fingerprint(key, hostname)
+  end
+
+  def user_key(_alg, opts) do
```

Comment:
This seems to completely disregard the actual algorithm passed.

According to the docs, the `Algorithm` parameter is there for:

> Host key algorithm. Is to support `'ssh-rsa'| 'ssh-dss'` but more algorithms can be handled.

If these need different handling, it may be better to explicitly add `user_key/2` functions which match on the supported algos, e.g. `def user_key('ssh-rsa', options) do …`.